File: src_cpp/ExpectimaxSearcher.cpp
#include "ExpectimaxSearcher.h"
#include <vector>
#include <iostream>

ExpectimaxSearcher::ExpectimaxSearcher() = default;

int ExpectimaxSearcher::find_best_move_with_eval(
    const std::array<std::array<int, 4>, 4>& board,
    int depth,
    const std::function<float(const std::array<std::array<int, 4>, 4>&)>& eval_func) {
    float best_score = -1e9;
    int best_move = -1;

    for (int move = 0; move < 4; ++move) {
        Fast2048 temp_game;
        temp_game.set_board(board);

        if (!temp_game.is_move_valid(move)) {
            continue;
        }

        auto [merge_score, done, moved] = temp_game.move(move);

        float score = static_cast<float>(merge_score) + chance_node(temp_game, depth, eval_func);
        if (score > best_score) {
            best_score = score;
            best_move = move;
        }
    }

    if (best_move == -1) {
        Fast2048 temp_game;
        temp_game.set_board(board);
        for(int m = 0; m < 4; ++m) {
            if (temp_game.is_move_valid(m)) return m;
        }
    }

    return best_move;
}

float ExpectimaxSearcher::chance_node(Fast2048& game, int depth, const std::function<float(const std::array<std::array<int, 4>, 4>&)>& eval_func) {
    std::vector<std::pair<int, int>> empty_cells;
    auto board = game.get_board();
    for (int r = 0; r < 4; ++r) {
        for (int c = 0; c < 4; ++c) {
            if (board[r][c] == 0) {
                empty_cells.emplace_back(r, c);
            }
        }
    }

    if (empty_cells.empty()) {
        return 0.0f;
    }

    float total_value = 0.0f;
    for (const auto& cell : empty_cells) {
        Fast2048 game_with_2 = game;
        auto board_with_2 = game_with_2.get_board();
        board_with_2[cell.first][cell.second] = 1;
        game_with_2.set_board(board_with_2);
        total_value += 0.9f * max_node(game_with_2, depth - 1, eval_func);

        Fast2048 game_with_4 = game;
        auto board_with_4 = game_with_4.get_board();
        board_with_4[cell.first][cell.second] = 2;
        game_with_4.set_board(board_with_4);
        total_value += 0.1f * max_node(game_with_4, depth - 1, eval_func);
    }

    // Return the average expected value
    return total_value / empty_cells.size();
}

float ExpectimaxSearcher::max_node(Fast2048& game, int depth, const std::function<float(const std::array<std::array<int, 4>, 4>&)>& eval_func) {
    if (depth <= 0) {
        return eval_func(game.get_board());
    }

    float max_value = -1e9;
    bool any_move_possible = false;

    for (int move = 0; move < 4; ++move) {
        Fast2048 temp_game = game;

        if (!temp_game.is_move_valid(move)) {
            continue;
        }

        auto [merge_score, done, moved] = temp_game.move(move);

        if (!moved) {
            continue;
        }
        
        any_move_possible = true;
        float value = static_cast<float>(merge_score) + chance_node(temp_game, depth, eval_func);
        if (value > max_value) {
            max_value = value;
        }
    }

    if (!any_move_possible) {
        return 0.0f;
    }

    return max_value;
}

File: src/ExpectimaxSearcher.py
import numpy as np
import torch
from .Fast2048 import Fast2048
from .utility import board_to_tensor


class ExpectimaxSearcher:
    def __init__(self, model, device="cuda"):
        self.model = model
        self.device = device
        self.transposition_table = {}

    def find_best_move(self, board, search_depth=3):
        best_score = -float('inf')
        best_move = -1

        self.transposition_table = {}

        for move in range(4):
            temp_game = Fast2048()
            temp_game.board = board.copy()

            if not temp_game.is_move_valid(move):
                continue

            merge_score, done,_  = temp_game.move(move)

            score = merge_score + self._chance_node(temp_game.board, search_depth)
            if score > best_score:
                best_score = score
                best_move = move

        return best_move if best_move != -1 else 0

    def _chance_node(self, board, depth):
        board_tuple = tuple(map(tuple, board))
        if (board_tuple, depth) in self.transposition_table:
            return self.transposition_table[(board_tuple, depth)]

        empty_cells = np.argwhere(board == 0)
        if len(empty_cells) == 0:
            return 0

        # Consider placing a '2' (log value 1) in each empty cell
        value_sum_2 = 0
        for r, c in empty_cells:
            temp_board = board.copy()
            temp_board[r, c] = 1  # log2(2) = 1
            value_sum_2 += self._max_node(temp_board, depth - 1)

        # Consider placing a '4' (log value 2) in each empty cell
        value_sum_4 = 0
        for r, c in empty_cells:
            temp_board = board.copy()
            temp_board[r, c] = 2  # log2(4) = 2
            value_sum_4 += self._max_node(temp_board, depth - 1)

        # Calculate the final probability-weighted average value
        expected_value = (0.9 * (value_sum_2 / len(empty_cells))) + \
                         (0.1 * (value_sum_4 / len(empty_cells)))

        # Store the result in the cache before returning
        self.transposition_table[(board_tuple, depth)] = expected_value
        return expected_value

    def _max_node(self, board, depth):
        board_tuple = tuple(map(tuple, board))
        if (board_tuple, depth) in self.transposition_table:
            return self.transposition_table[(board_tuple, depth)]

        if depth == 0:
            obs = board_to_tensor(board)
            with torch.no_grad():
                value = self.model.policy.predict_values(
                    torch.as_tensor(obs[None]).to(self.device)
                )
            return value.item()

        max_value = -float('inf')

        for move in range(4):
            temp_game = Fast2048()
            temp_game.board = board.copy()
            if not temp_game.is_move_valid(move):
                continue
            merge_score, done, moved = temp_game.move(move)

            value = merge_score + self._chance_node(temp_game.board, depth)
            if value > max_value:
                max_value = value

        final_value = max_value if max_value != -float('inf') else 0
        self.transposition_table[(board_tuple, depth)] = final_value
        return final_value
